packages/datadog-plugin-jest/src/index.js

```diff
@@ -0,0 +1,197 @@
+      [TEST_SUITE]: testSuite
+    })
+}
+
+module.exports = function (BaseEnvironment) {
+  return class DatadogJestEnvironment extends BaseEnvironment {
+    constructor (config, context) {
+      super(config, context)
+      this.testSuite = context.testPath.replace(`${config.rootDir}/`, '')
+      this.rootDir = config.rootDir
+    }
+    async setup () {
+      if (!global.tracer) {
+        const ciMetadata = getCIMetadata()
+        const { repository, branch, commit } = await getGitInformation()
+        global.tracer = require('../../dd-trace').init({
+          sampleRate: 1,
+          flushInterval: 1,
+          startupLogs: false,
+          ingestion: {
+            sampleRate: 1
+          },
+          tags: {
+            ...ciMetadata,
+            [GIT_COMMIT_SHA]: commit,
+            [GIT_BRANCH]: branch,
+            [GIT_REPOSITORY_URL]: repository,
+            [BUILD_SOURCE_ROOT]: this.rootDir,
+            [TEST_FRAMEWORK]: 'jest'
+          }
+        })
+      }
+
+      return super.setup()
+    }
+    async teardown () {
+      await new Promise((resolve) => {
+        global.tracer._tracer._exporter._writer.flush(resolve)
+      })
+      return super.teardown()
+    }
+
+    async handleTestEvent (event) {
+      if (event.name === 'test_skip' || event.name === 'test_todo') {
+        global.tracer.trace(
+          'jest.test',
+          { type: 'test', resource: `${event.test.parent.name}.${event.test.name}` },
+          (span) => {
+            span.addTags({
+              [TEST_TYPE]: 'test',
+              [TEST_NAME]: event.test.name,
+              [TEST_SUITE]: this.testSuite,
+              [TEST_STATUS]: 'skip'
+            })
+          }
+        )
+      }
+      if (event.name === 'test_start') {
+        const originalSpecFunction = event.test.fn
+        if (originalSpecFunction.length) {
+          event.test.fn = global.tracer.wrap(
+            'jest.test',
+            { type: 'test', resource: `${event.test.parent.name}.${event.test.name}` },
+            () => {
+              addTestMetadata({ testName: event.test.name, testSuite: this.testSuite })
+              return new Promise((resolve, reject) => {
+                originalSpecFunction((err) => {
+                  if (err) {
+                    setTestStatus('fail')
+                    reject(err)
+                  } else {
+                    setTestStatus('pass')
+                    resolve()
+                  }
+                  finishStartedSpans()
+                })
+              })
+            }
+          )
+        } else {
+          event.test.fn = global.tracer.wrap(
+            'jest.test',
+            { type: 'test', resource: `${event.test.parent.name}.${event.test.name}` },
+            () => {
+              let result
+              addTestMetadata({ testName: event.test.name, testSuite: this.testSuite })
+              try {
+                result = originalSpecFunction()
```

Comment:
I think you can avoid all the promise-handling code below by just making this function (i.e. the function starting on line 166) an async function and awaiting here on line 170. Synchronous functions can be awaited, so this should work just fine.